chatbot-core/llm.py

```diff
@@ -0,0 +1,177 @@
+        """Get the prompt template for response generation"""
+        system_prompt = self._system_prompt_for_response()
+        llm_response_prompt = ChatPromptTemplate.from_messages(
+            [("system", system_prompt), MessagesPlaceholder("chat_history"), ("human", "{input}")]
+        )
+        return llm_response_prompt
+
+    def get_db_retriever(
+        self, collection_name: str, embedding_model: Any, top_k_items: int = 5, score_threshold: float = 0.5
+    ):
+        """Get the database retriever"""
+        db_retriever = QdrantVectorStore(
+            client=self.qdrant_client, collection_name=collection_name, embedding=embedding_model
+        )
+        retriever = db_retriever.as_retriever(
+            search_type="similarity_score_threshold", search_kwargs={"k": top_k_items, "score_threshold": score_threshold}
+        )
+        return retriever
+
+    def create_chain(self, db_collection_name: str):
+        """Creates a llm chain"""
+        if not self.llm_model:
+            raise Exception("The LLM model is not loaded.")
+
+        embedding_model = self.get_selected_embedding_model(EMBEDDING_MODEL_CHOICE)
+        context_prompt_template = self.get_prompt_template_for_retrieval()
+        response_prompt_template = self.get_prompt_template_for_response()
+
+        retriever = self.get_db_retriever(collection_name=db_collection_name, embedding_model=embedding_model)
+
+        history_aware_retriever = create_history_aware_retriever(self.llm_model, retriever, context_prompt_template)
+
+        chat_response_chain = create_stuff_documents_chain(self.llm_model, response_prompt_template)
+
+        rag_chain = create_retrieval_chain(history_aware_retriever, chat_response_chain)
+        return rag_chain
+
+    def execute_chain(self, query: str, db_collection_name: str = QDRANT_DB_COLLECTION_NAME):
+        """
+        Executes the chain
+        """
+        rag_chain = self.create_chain(db_collection_name=db_collection_name)
+
+        response = rag_chain.invoke({"input": query, "chat_history": self.get_message_history()["chat_history"]})
+        self.memory.chat_memory.add_message(HumanMessage(content=query))
+        self.memory.chat_memory.add_message(AIMessage(content=response["answer"]))
+        return response["answer"] if "answer" in response else ""
+
+    def get_message_history(self):
+        """
+        Returns the historical conversational data
+        """
+        return self.memory.load_memory_variables({})
+
+
+@dataclass
+class OpenAIHandler(LLMBase):
+    """LLM handler using OpenAI for RAG"""
+
+    model: str = "gpt-3.5-turbo-1106"
+    temperature: float = 0.2
+
+    def __post_init__(self):
+        super().__post_init__()
+
+        self.llm_model = ChatOpenAI(model=self.model, temperature=self.temperature)
+
+
+@dataclass
+class OllamaHandler(LLMBase):
+    """LLM Handler using Ollama for RAG"""
+
+    model: str = "mistral:latest"
+    base_url: str = "http://localhost:11434"
```

Comment:
Let's not use default values.

```suggestion
    base_url: str
```